src/libpcp_web/src/timer.c

```diff
@@ -0,0 +1,222 @@
+    usr *= 1000.0; /* milliseconds */
+    sys *= 1000.0;
+    if (getrusage(RUSAGE_SELF, &usage) < 0)
+	return;
+
+    if ((value = mmv_lookup_value_desc(map, "cpu.user", NULL)) != NULL)
+	mmv_set_value(map, value, usr);
+    if ((value = mmv_lookup_value_desc(map, "cpu.sys", NULL)) != NULL)
+	mmv_set_value(map, value, sys);
+    if ((value = mmv_lookup_value_desc(map, "cpu.total", NULL)) != NULL)
+	mmv_set_value(map, value, usr+sys);
+    if ((value = mmv_lookup_value_desc(map, "mem.maxrss", NULL)) != NULL)
+	mmv_set_value(map, value, usage.ru_maxrss);
+    if ((value = mmv_lookup_value_desc(map, "mem.datasz", NULL)) != NULL)
+	mmv_set_value(map, value, (double)datasz);
+}
+
+/*
+ * Register and set up generic server metrics.
+ * See pmproxy/src/server.c for calling example.
+ */
+int
+pmWebTimerSetMetricRegistry(struct mmv_registry *registry)
+{
+
+    pmAtomValue		*value;
+    pmInDom		noindom = MMV_INDOM_NULL;
+    pmUnits		nounits = MMV_UNITS(0,0,0,0,0,0);
+    pmUnits		units_kbytes = MMV_UNITS(1, 0, 0, PM_SPACE_KBYTE, 0, 0);
+    pmUnits		units_msec = MMV_UNITS(0, 1, 0, 0, PM_TIME_MSEC, 0);
+    pid_t		pid = getpid();
+    char		buffer[64];
+
+    if (server_registry) {
+	fprintf(stderr, "%s: Error: server instrumentation already registered\n", pmGetProgname());
+    	return -EINVAL;
+    }
+
+    server_registry = registry;
+    mmv_stats_add_metric(registry, "pid", SERVER_PID,
+		MMV_TYPE_U32, MMV_SEM_DISCRETE, nounits, noindom,
+		"server PID",
+		"PID for the current server invocation");
+    pmsprintf(buffer, sizeof(buffer), "%u", pid);
+    mmv_stats_add_metric_label(registry, SERVER_PID,
+		"pid", buffer, MMV_NUMBER_TYPE, 0);
+
+    mmv_stats_add_metric(registry, "cpu.user", SERVER_CPU_USR,
+	MMV_TYPE_U64, MMV_SEM_COUNTER, units_msec, noindom,
+	"server user CPU",
+	"server process user CPU time counter");
+
+    mmv_stats_add_metric(registry, "cpu.sys", SERVER_CPU_SYS,
+	MMV_TYPE_U64, MMV_SEM_COUNTER, units_msec, noindom,
+	"server system CPU",
+	"server process system CPU time counter");
+
+    mmv_stats_add_metric(registry, "cpu.total", SERVER_CPU_TOT,
+	MMV_TYPE_U64, MMV_SEM_COUNTER, units_msec, noindom,
+	"server system + user CPU",
+	"server process system + user CPU time");
+
+    mmv_stats_add_metric(registry, "mem.maxrss", SERVER_MEM_MAXRSS,
+	MMV_TYPE_U64, MMV_SEM_INSTANT, units_kbytes, noindom,
+	"server maximum RSS",
+	"server process maximum resident set memory size");
+
+    mmv_stats_add_metric(registry, "mem.datasz", SERVER_MEM_DATASZ,
+	MMV_TYPE_U64, MMV_SEM_INSTANT, units_kbytes, noindom,
+	"server virtual data size",
+	"server process data memory size, returned from sbrk(2)");
+
+    if ((server_map = mmv_stats_start(registry)) == NULL) {
+	fprintf(stderr, "%s: Error: server instrumentation disabled\n", pmGetProgname());
+	return -EINVAL;
+    }
+
+    /* PID doesn't change, register it once */
+    if ((value = mmv_lookup_value_desc(server_map, "pid", NULL)) != NULL)
+	mmv_set_value(server_map, value, pid);
+
```

Comment:
About here is where we should be looking up all the metric offsets and stashing them in an array for subsequent direct lookup (as described earlier).